### PR TITLE
Fix some clippy lints

### PR DESCRIPTION
**Pull request by alanhdu**
I went ahead and fixed all the clippy lints (there were a couple I thought would be better unfixed and added `cfg` statements to allow them) and also re-enabled clippy and rustfmt in CI.

They were disabled with `no time to explain, disable clippy checks`, so hopefully this won't go against whatever the reason at the time was :laughing:.

One question about the CI though: right now, it's an allowed failure that runs against the latest nightly each time. Would it be better to pin it to a specific nightly (or use the `beta` versions) to lower the churn?

### CODE REVIEW

**Comment by matklad** [comment]
>They were disabled with no time to explain, disable clippy checks, so hopefully this won't go against whatever the reason at the time was laughing.

At that time, I was transitioning to the to the new implementation of the syntax tree, so that code churn rate was high enough for clippy & rustfmt being a significant problem.


Overall, I am torn about this, curious what other folks think. Here are my rough thoughts about the issue.

I definitely don't want to gate PRs on clippy and rustfmt: to me, additional per-PR code improvements that these tools bring feel much less important that annoyance and additional waiting time for the PR.

I do want to use rustfmt and, to a lesser extend, clippy. These tools do improve code quality in minor-to-moderate ways. At the same time, I am not too keen on keeping a clean git history, especially since GitHub implemented the feature to get the previous blame. So applying suggestions manually from time to time seems fine by me.

I am annoyed by some of the clippy suggestions, which make the code worse :) In particular:

* `needless_pass_by_value` is often very intentional and important for correctness. For example, when passing `Sender`s around the moment when the last `Sender` is dropped is important, because it tears down the receiver and allows to wait for it. 

* Minor changes to boolean expressions like `!(idx < container.len())` to `idx >= container.len()` could make the code less readable. In this case, `idx < container.len()` is the usual invariant, and `!` very cleary states "invariant does not hold". 

* The `.map(|&it| it)` to `.cloned()` changes obscure the fact that it is almost-zero const operation, because the type is `Copy`. 


I am also not thrilled by the idea of disabling clippy lints on a case-by-case bases. Is there perhaps some kind of conservative unopinionated clippy profile, that could be applied on the project level?

**Comment by DJMcNab** [comment]
I think we should enforce rustfmt in ci, but then just run clippy every so often if we find a good set of options.

**Comment by matklad** [comment]
>I think we should enforce rustfmt in ci, 

@DJMcNab why would be more useful than, say, running rusfmt once a week and adding

```
add this to .git/hooks/pre-commit to run rustfmt automatically on save

cargo fmt-changes
```

where fmt-changes is a command which applies rustfmt to changed files?

I think it's a bad contributor experience when you PR fails just because you've forgotten a trailing comma, or use a different version of rustfmt.

**Comment by DJMcNab** [comment]
Different rustfmt version is a good point. I guess I've been spoiled by the ecosystem around nodejs, especially prettier combined with husky and vscode that I think of formatting as a non-issue.

The Rust formatting ecosystem isn't quite there yet, unfortunately, so maybe it would be better to format every so often. I wonder if bors can handle rustfmting automatically somehow...

**Comment by CAD97** [comment]
You can (but shouldn't) make CI commit and push to the repo.

Maybe someone could set up a bot that does fmt PRs?

(Personally, I enforce a pinned stable rustfmt. I think it's better to require standard formatting always, which keeps git-blame accurate. The CI job is named so it's simple to see why it failed.)

**Comment by alanhdu** [comment]
Definitely here you about some of the lints -- I've rebased this branch and only cherry-picked the Clippy lints that I think are probably unobjectionable.

>  Is there perhaps some kind of conservative unopinionated clippy profile, that could be applied on the project level?

I think so, but it looks like there's no way to configure it on a per-repo level and you need to add `[cfg]` statements per-crate -- surprisingly, it doesn't look like there's a way to do per-repo configuration via a toml file somewhere (at least based on https://github.com/rust-lang/cargo/issues/5034 and https://github.com/rust-lang-nursery/rust-clippy/issues/1313).

I think not-having clippy in CI and only doing it "once in a while" is probably a good choice until lints can be configured centrally and not per-crate. I think it's reasonable to do `rustfmt` in CI, but don't feel strongly (for the Rust  projects I work on, we have a "blessed" version of Rust to develop on, and we use the corresponding `rustfmt` that's `rustup` distributes with that version).

**Comment by alanhdu** [comment]
I guess we could also configure clippy through the command line:

```
cargo +nightly clippy -- -A clippy::style -A clippy::nonminimal_bool
```

seems to work fine.

File: crates/ra_editor/src/scope/fn_scope.rs [hunk 1]
```diff
@@ -88,7 +88,7 @@ pub struct ScopeEntry {
 }
 
 impl ScopeEntry {
-    fn new(pat: ast::BindPat) -> Option<ScopeEntry> {
+    fn new_opt(pat: ast::BindPat) -> Option<ScopeEntry> {
```
Comment (matklad) [inline crates/ra_editor/src/scope/fn_scope.rs]: Let's use `new` to name such methods? It's what stdlib does, so it looks like a bug in clippy:

https://doc.rust-lang.org/std/ffi/struct.CString.html#method.new

File: crates/ra_analysis/src/job.rs [hunk 2]
```diff
@@ -11,7 +11,7 @@ pub struct JobToken {
 }
 
 impl JobHandle {
-    pub fn new() -> (JobHandle, JobToken) {
+    pub fn new_pair() -> (JobHandle, JobToken) {
```
Comment (matklad) [inline crates/ra_analysis/src/job.rs]: Agree that `new` probably isn't the best name here, the whole `JobHandle` infra will be removed once we finish migration to salsa.

File: crates/ra_editor/src/scope/fn_scope.rs (hunk 1 above)
Comment (alanhdu) [reply crates/ra_editor/src/scope/fn_scope.rs]: Looks like this will be fixed with https://github.com/rust-lang-nursery/rust-clippy/pull/3338

**Comment by matklad** [comment]
bors r+

Thanks!

**Comment by matklad** [comment]
Filed https://github.com/rust-analyzer/rust-analyzer/issues/155 for rustfmt

**Pull request merged**